benchmarks/cpp/transformer.cpp

```diff
@@ -0,0 +1,228 @@
+#include <tests/cpp/multidevice.h>
+#include <tests/cpp/multidevice_transformer.h>
+#include <tests/cpp/utils.h>
+#include <utils.h>
+
+using namespace nvfuser;
+
+constexpr int64_t B = 1, E = 12288, H = 96, S = 2048;
+constexpr double kParamScale = 0.02;
+constexpr double kDropoutProb = 0.0, kSdpaProb = 0.0;
+constexpr int64_t warmup_itrs = 10, num_itrs = 10;
+
+namespace {
+at::Tensor shardTensor(
+    at::Tensor tensor,
+    const int64_t axis,
+    const DeviceMesh& mesh,
+    Communicator* communicator) {
+  const auto device_id = communicator->deviceId();
+  return nvfuser::shardTensor(tensor, axis, mesh, device_id);
+}
+} // namespace
+
+void forward_transformer(Communicator* communicator, bool profile, bool sequence_parallel) {
+  int64_t D = communicator->size();
+  if (D == 1) {
+    std::cout << "Sequence parallel requires >1 devices, D=" << D << std::endl;
+  }
+  auto dtype = DataType::BFloat16;
+  at::ScalarType at_dtype = data_type_to_aten(dtype);
+  const auto mesh = DeviceMesh::createForNumDevices(D);
+  const auto options =
+      at::TensorOptions().dtype(at_dtype).device(communicator->device());
+
+  auto x = at::randn({B * S, E}, options);
+  auto ln0_w = at::randn(E, options).to(at::kFloat);
+  auto ln0_b = at::randn(E, options).to(at::kFloat);
+  auto mha_w0 = at::randn({3 * E, E}, options) * kParamScale;
+  auto mha_b0 = at::randn({3 * E}, options) * kParamScale;
+  auto mha_w1 = at::randn({E, E}, options) * kParamScale;
+  auto mha_b1 = at::randn({E}, options) * kParamScale;
+  auto ln1_w = at::randn(E, options).to(at::kFloat);
+  auto ln1_b = at::randn(E, options).to(at::kFloat);
+  auto mlp_w0 = at::randn({4 * E, E}, options) * kParamScale;
+  auto mlp_b0 = at::randn({4 * E}, options) * kParamScale;
+  auto mlp_w1 = at::randn({E, 4 * E}, options) * kParamScale;
+  auto mlp_b1 = at::randn({E}, options) * kParamScale;
+
+  std::vector<c10::IValue> at_inputs = {
+      sequence_parallel ? shardTensor(x, 0, mesh, communicator).unsqueeze(0) : x,
+      ln0_w,
+      ln0_b,
+      shardTensor(mha_w0.view({3, E, E}), 1, mesh, communicator)
+          .view({1, 3 * E / D, E}),
+      shardTensor(mha_b0.view({3, E}), 1, mesh, communicator)
+          .view({1, 3 * E / D}),
+      shardTensor(mha_w1, 1, mesh, communicator).unsqueeze(0),
+      mha_b1,
+      ln1_w,
+      ln1_b,
+      shardTensor(mlp_w0, 0, mesh, communicator).unsqueeze(0),
+      shardTensor(mlp_b0, 0, mesh, communicator).unsqueeze(0),
+      shardTensor(mlp_w1, 1, mesh, communicator).unsqueeze(0),
+      mlp_b1};
+
+  DistributedTransformer model = DistributedTransformer(D, B, E, H, S, kDropoutProb, kSdpaProb);
+  auto fec = model.forward(dtype, sequence_parallel);
+  cudaSetDevice(communicator->deviceId());
+
+  auto start = std::chrono::high_resolution_clock::now();
+  for (auto i : c10::irange(num_itrs + warmup_itrs)) {
+    if (i == warmup_itrs) {
+      start = std::chrono::high_resolution_clock::now();
+      if (profile) {
+        cudaProfilerStart();
```

Comment:
It is probably safer to profile and measure time in separate iterations. IOW, to have different phases of iterations, like

```
warmup1
nsight profiling
warmup2
chrono
```